devito/operator/operator.py

```diff
@@ -754,11 +761,12 @@ def _soname(self):
 
     @cached_property
     def ccode(self):
-        try:
-            return self._ccode_handler(compiler=self._compiler).visit(self)
-        except (AttributeError, TypeError):
-            from devito.ir.iet.visitors import CGen
-            return CGen(compiler=self._compiler).visit(self)
+        with switchconfig(compiler=self._compiler, language=self._language):
```

Comment:
changing the `configuration` state on such a scale is a bit evil IMHO, I'd rather propagate the compiler/language down as necessary.

Also a minor but maybe important question: if for example we're writing a test, something along the lines of

```
try:
     op(...)
except:
     ...
```

or simply

```
with pytest.raises(...):
    op(...)
 ```
 
and an exception is raised within `op()`, isn't there a risk that the `configuration` state is changed and never restored?

Reply:
The except should still act as an exit, it doesn't kill python so it shouldn't be an issue.

So you would prefer to have `ccode` take the language as input? That's fine too, but this still makes sure that no matter what, the configuration is the correct one for the entire visitor pass.